transport.go

```diff
@@ -164,14 +177,22 @@ func (t *Transport) refreshToken(ctx context.Context) error {
 	t.appsTransport.BaseURL = t.BaseURL
 	t.appsTransport.Client = t.Client
 	resp, err := t.appsTransport.RoundTrip(req)
+	e := &RefreshTokenError{
+		RootCause:      err,
+		InstallationID: t.installationID,
+		Response:       resp,
+	}
 	if err != nil {
-		return fmt.Errorf("could not get access_tokens from GitHub API for installation ID %v: %v", t.installationID, err)
+		e.Message = fmt.Sprintf("could not get access_tokens from GitHub API for installation ID %v: %v", t.installationID, err)
+		return e
 	}
-	defer resp.Body.Close()
 
 	if resp.StatusCode/100 != 2 {
-		return fmt.Errorf("received non 2xx response status %q when fetching %v", resp.Status, req.URL)
+		e.Message = fmt.Sprintf("received non 2xx response status %q when fetching %v", resp.Status, req.URL)
+		return e
 	}
+	// Closing body late, to provide caller a chance to inspect body in an error / non-200 response status situation
+	defer resp.Body.Close()
```

Comment:
Since defer is scoped to the func, this is going to close the body before any callers can look at it. We'll need to copy the bytes into a separate `io.ReadCloser` (`io.Copy` + `bytes.Buffer` + `io.NopCloser` will probably work well here).

https://go.googlesource.com/go/+/go1.16/src/net/http/httputil/dump.go might be a helpful as an existing example.

Reply:
Not sure if I can follow your thoughts.
We call the `defer` only in case no HTTPError is generated. The following `Decode` can still throw an error.
Do you aim to provide the body for inspection every time, even if no error is thrown?

Reply:
You're right! Ignore this.